lib/Transforms/LLZKInlineStructsPass.cpp

```diff
@@ -0,0 +1,555 @@
+    }
+
+    MLIRContext *ctx = destStruct.getContext();
+    RewritePatternSet patterns(ctx);
+    patterns.add<EraseExcessCreateStructOp>(ctx, srcStruct.getType());
+    // Just run on "compute()" function because `CreateStructOp` cannot be used in "constrain()"
+    walkAndApplyPatterns(destStruct.getComputeFuncOp(), std::move(patterns));
+
+    return success();
+  }
+
+public:
+  StructInliner(SymbolTableCollection &tables, StructDefOp from, StructDefOp into)
+      : StructInlinerBase(tables), srcStruct(from), destStruct(into) {}
+
+  LogicalResult doInline() {
+    LLVM_DEBUG(
+        llvm::dbgs() << "[StructInliner] merge " << srcStruct.getSymNameAttr() << " into "
+                     << destStruct.getSymNameAttr() << '\n'
+    );
+
+    DestToSrcToClonedSrcInDest destToSrcToClone = cloneFields();
+    if (failed(handleConstrainCall(destToSrcToClone))) {
+      return failure();
+    }
+    if (failed(handleComputeCall(destToSrcToClone))) {
+      return failure();
+    }
+    if (failed(deleteFields(destToSrcToClone))) {
+      return failure();
+    }
+    return handleRemainingStructValues();
+  }
+};
+
+class InlineStructsPass : public llzk::impl::InlineStructsPassBase<InlineStructsPass> {
+  /// Maps caller struct to callees that should be inlined. The outer SmallVector preserves the
+  /// ordering from the bottom-up traversal that builds the InliningPlan so performing inlining
+  /// in the order given will not lose any or require doing any more than once.
+  // TODO: however, applying in the opposite direction would reduce making clones of clones of ...
+  // for the ops within the inlined struct functions, as they are inlined further up the tree. But
+  // would have to make sure it's done properly, and update some mappings in this plan as we go.
+  using InliningPlan = SmallVector<std::pair<StructDefOp, SmallVector<StructDefOp>>>;
+
+  static unsigned complexity(FuncDefOp f) {
+    unsigned complexity = 0;
+    f.getBody().walk([&complexity](Operation *op) {
+      // TODO: `EmitEqualityOp` and `EmitContainmentOp` should probably increment based on
+      // dimension sizes in the operands rather than just +1 in all cases.
+      if (llvm::isa<constrain::EmitEqualityOp, constrain::EmitContainmentOp, felt::MulFeltOp>(op)) {
+        ++complexity;
+      }
+    });
+    return complexity;
+  }
+
+  static FailureOr<FuncDefOp>
+  getIfStructConstrain(const SymbolUseGraphNode *node, SymbolTableCollection &tables) {
+    auto lookupRes = node->lookupSymbol(tables, false);
+    assert(succeeded(lookupRes) && "graph contains node with invalid path");
+    if (auto f = llvm::dyn_cast<FuncDefOp>(lookupRes->get())) {
+      if (f.isStructConstrain()) {
+        return f;
+      }
+    }
+    return failure();
+  }
+
+  /// Return the parent StructDefOp for the given Function (which is known to be a struct
+  /// "constrain" function so it must have a StructDefOp parent).
+  static inline StructDefOp getParentStruct(FuncDefOp func) {
```

Comment:
I think this should also be moved to some kind of utilities collection.

Reply:
The first assert is specific to this usage and the second would not be used in the common impl because it would probably return the `FailureOr` result itself. Without that, it's just `getParentOfType<StructDefOp>(..)` so I think no need to add a utility.